bin/warnhist

```diff
@@ -67,29 +86,196 @@ sub do_help {
+		    $fort_files{$bn} = [ $name ];
+		} else {
+		    push @{ $fort_files{$bn} }, $name;
+		}
+            # C++ source file
+	    } elsif($bn =~ /.*(\.cpp)|(\.c\+\+)|(\.cc)|(\.cp)|(\.cxx)$/i) {
+		$bn =~ s/(\.cpp)|(\.c\+\+)|(\.cc)|(\.cp)|(\.cxx)$//ig;
+		if(!exists $cpp_files{$bn}) {
+		    $cpp_files{$bn} = [ $name ];
+		} else {
+		    push @{ $cpp_files{$bn} }, $name;
+		}
+            # C source file
+	    } elsif($bn =~ /.*(\.c)$/i) {
+		$bn =~ s/(\.c)$//g;
+		if(!exists $c_files{$bn}) {
+		    $c_files{$bn} = [ $name ];
+		} else {
+		    push @{ $c_files{$bn} }, $name;
+		}
+	    }
+      }
+    };
+    find($path_checker, $root_path);
+
+#print Dumper \%c_files;
+#print Dumper \%cpp_files;
+#print Dumper \%fort_files;
+}
+
+sub count_file_loc {
+    my ($filename, $typename, $file_paths, $files_counted, $count) = @_;
+
+    # Attempt to detect and handle object file name mangling by Automake
+    if(!exists $file_paths->{$filename} && $filename =~ /\-/) {
+        my ($bn) = $filename =~ /\S+\-(\S+)$/x;
+        if(exists $file_paths->{$bn}) {
+            if(!exists $options{q}) {
+                warn "No path for $typename source file '$filename', but '$bn' has path, assuming automake generated object file name";
+            }
+            $filename = $bn;
+        }
+    }
+
+    if(exists $file_paths->{$filename}) {
+        my $filecount = 0;
+
+        # Attempt to count LOC for files with same name
+        if(scalar(@{$file_paths->{$filename}}) > 1) {
+            $filecount = $files_counted->{$filename}++;
+
+            # Issue warning about multiple source files with same name
+            if($filecount == 0 && !exists $options{q}) {
+                local $" = ', ';        # '$"' is documented in https://perldoc.perl.org/perlvar
+                warn "Multiple paths for $typename source file named '$filename', assuming each is compiled once, paths: [@{$file_paths->{$filename}}]\n";
+            }
+
+            # Sanity check for too many compiles of a file
+            if($filecount >= scalar(@{$file_paths->{$filename}})) {
+                if(!exists $options{q}) {
+                    local $" = ', ';        # '$"' is documented in https://perldoc.perl.org/perlvar
+                    warn "Too many compiles of $typename source file named '$filename' with paths: [@{$file_paths->{$filename}}], disabling warning density calculations\n";
+                }
+                delete $options{d};
+            }
+        }
+
+        # Increment the # of lines of code (if not too many)
+        if($filecount < scalar(@{$file_paths->{$filename}})) {
+            ${$count} += line_count($file_paths->{$filename}[$filecount]);
+        }
+    } else {
+        if(!exists $options{q}) {
+            warn "No path for $typename source file '$filename', e '-p' option to specify, disabling warning density calculations\n";
+        }
+        delete $options{d};
+    }
+}
+
+# Compute LOC for compiled source file
```

Comment:
Determine the type of source file being compiled, then locate it in the source directory hierarchy(s), and count its lines of code.